### ISSUE TITLE
could not read AUTH username when using AIDL interface to start/import profile

### ISSUE THREAD

**Issue by tr56e** [schwabe/ics-openvpn#1342]
1. Android One / 10 
2. Android One / build 00WW_2_410_SP03
3. Nokia 7.2
4. 0.7.22
 
# Description of the issue

we use the app with a managed configuration. Occasionally I get the error "could not read AUTH username / password / ok / string from management interface" and after restarting the device, the user name and password are requested and not saved.
# General information
MDM: Sophos Mobile

**Comment by schwabe** [comment]
I need a bit more information like the exact error message to understand what is happening here.

**Comment by yoopernc** [comment]
I have a similar situation. I'm using a Samsung S20 running Android 10 with de.blinkt OpenVPN ver 0.6.66. When i pass my config to OpenVPN, I'm prompted for a Username & Password. After entering this, OpenVPN's status becomes "Building configuration...", and it sits. Reading the adb logcat, I see:

I openvpn : Got unrecognized command>FATAL:ERROR: could not read Auth username/password/ok/string from management interface
I OpenVPN : OpenVPN process exited
I OpenVPN : Exiting

About every minute I'm prompted for the UID/PW again, and the cycle repeats.

Edit: I should mention that I'm doing all of this using a Device Owner DPC program. In this, I receive status updates from OpenVPN. When I use certificates and a private key, everything works fine. But now I'm trying to authenticate using UserID/Password and no cert/key. Each time that I enter the UID/PW, I get a status update from OpenVPN with a valid UUID, state = VPN_GENERATE_CONFIG, message = "", and level = "LEVEL_START". At this point, I do not know what OpenVPN is expecting, because starting OpenVPN as a stand-alone app and feeding it the config (below) is enough to connect to the VPN using just a UID/PW for authentication.

Here's my config:

client
dev tun
persist-tun
persist-key
proto udp
remote vpn.xxxxxxxx.xxx nnnn
nobind
cipher AES-256-CBC
auth-user-pass
resolv-retry infinite
setenv CLIENT_CERT 0
redirect-gateway def1
dhcp-option DNS 8.8.8.8
auth RSA-SHA256
tls-client
remote-cert-tls server
key-direction 1
&lt;tls-auth&gt;
-----BEGIN OpenVPN Static key V1-----
206692e11cd...
...
...7a2c41c
-----END OpenVPN Static key V1-----
&lt;/tls-auth&gt;
&lt;ca&gt;
-----BEGIN CERTIFICATE-----
MIIFCTCCAvGg...
...
...ExqaUYA6238R
-----END CERTIFICATE-----
&lt;/ca&gt;

If you need any more information, I'll recreate it and get it to you. Thanks!

**Comment by yoopernc** [comment]
Further information:

When my process finally times out, after several retries, I'm able to pull down the notifications on my S20's screen. When I tap on the notification that says "OpenVPN... Building configuration...", it brings up the following screen:

<img width="417" alt="Screen Shot 2021-09-25 at 10 27 31 AM" src="https://user-images.githubusercontent.com/8331528/134775033-a3fd40f9-379e-4987-8553-56be6bf8729f.png">

My searches using this error message have not produced any useful results.

**Comment by yoopernc** [comment]
One more post with more info on my situation:

I'm trying to use OpenVPN API in my Android app. I'm calling:

- prepare
- prepareVPNService
- registerStatusCallback
- addNewVPNProfile (with the config above)
- startProfile, passing in the vpnProfile.mUUID from addNewVPNProfile

Instead of startProfile, I've tried calling "startVPN" with the config above, but that didn't work any better.

I'm hoping that you can decipher all of this and point me in the direction I need to go. Thanks!

**Comment by schwabe** [comment]
hm, that sounds really like a bug. For startprofile the username password should be embedded as 
```
<auth-user-pass>
user
pass
</auth-user-pass>
```
since startProfile only uses a temporary profile and so user experience suffer since save user/pass etc. doesn't work. But when added as normal profile it should work....

**Comment by MasterProvider** [comment]
Yes it will work in normal profile but with additional configuration will return to exit value 1

**Comment by MasterProvider** [comment]
I think the problem is the OpenVPN 2.x because when I use OpenVPN 3 Core It's working

**Comment by MasterProvider** [comment]
Yes it will work in normal profile but with additional configuration will return to exit value 1

**Comment by MasterProvider** [comment]
Yes it works in normal profile but when you put additional configuration will return to exit value 1

**Comment by schwabe** [comment]
@MasterProvider are you the same person as @yoopernc? I am a bit confused what you mean with "additional configuration" in this context

**Comment by MasterProvider** [comment]
@schwabe no I'm not but I have same issue with "Process  exited with exit value 1"

**Comment by MasterProvider** [comment]
Sorry for misunderstanding about "additional configuration" I mean custom option

**Comment by schwabe** [comment]
@MasterProvider sorry but same symptom does not mean same bug. Please keep that discussion outside of this bug because it confuses things.

**Comment by yoopernc** [comment]
@schwabe, I am not the same person as @MasterProvider or @tr56e. I saw the OP, and since my problem is similar, I posted here. I probably should have opened my own thread. I'm sorry about that.

I don't know how to embed the user/pass because I send the configuration to OpenVPN using addNewProfile, and then call startProfile. After this OpenVPN prompts me for the username and password. I would have assumed that OpenVPN would insert these values into the connection request that is sent to the VPN server.

My android app can connect successfully to my VPN host using OpenVPN if I include `<cert>...</cert>` and `<key>...</key>`. But I'm trying to authenticate using only username and password and the configuration I posted above. Again, this works if I start OpenVPN as a stand-alone app and pass in the configuration. It prompts for user/pass and connects successfully. But it fails when I try to do the same thing programmatically using the OpenVPN API calls.

**Comment by schwabe** [comment]
I edited my comment and yours to becuase github hides the `<key>` directives thinking you want to write html. My comment should now show you how to embed them.

**Comment by yoopernc** [comment]
Thank you for your reply. What you are saying is that I should get the Username and Password from the user myself and embed them into the configuration. I can test that by hard-coding them into the configuration. But can you tell me why letting OpenVPN prompt for the Username and Password does not work?

**Comment by schwabe** [comment]
for `startProfile` you need to embed them. For adding profile that is a bug that I need to look into. Those profile should behave exactly like a profile that a user added themselves.

**Comment by yoopernc** [comment]
I understand. Thank you.

What about the startVPN option? Should that work just the same as running OpenVPN as a stand-alone app and giving it the configuration that contains "auth-user-pass", then prompting for the two values?

**Comment by schwabe** [comment]
@yoopernc what version of ics-openvpn are you using? I just looked at this bug report again and noticed that it says `OpenSSL 1.0.2k` from 2017 in the log screen. That sounds like a four year old version.

**Comment by yoopernc** [comment]
@schwabe We are currently using version 0.6.66. I know it's four years old, but our customer would have to approve an upgrade, so there would have to be a good reason to ask them to do that. I did try version 0.7.24 to see if the problem was fixed, but it behaves the same way.

**Comment by schwabe** [comment]
I tried to reproduce the problem. But at least when I use `startVPN` or the `addNewVPNProfile`/`startProfile` combo with an inline configuration that has something like
```<auth-user-pass>
username
password
</auth-user-pass>
```
in it, it starts just fine.

There was a race condition with the username/password if just `auth-user-pass` was used instead.

**Comment by yoopernc** [comment]
Thank you for your summary. I am able to make a connection by appending this to the VPN Profile:

```
<auth-user-pass>
username
password
</auth-user-pass>
```

So, for now I am presenting the user with my own Username/Password entry dialog and then creating the Vpn Profile with that addition. Also, I am removing `auth-user-pass` from the original Profile. Is this necessary? If not, I can simplify my code a little by leaving it in.

When do you plan on making a stable release with this fix?

**Comment by schwabe** [comment]
This is mainly a spare time project, so I don't give any ETA on anything. The parser probably takes the second auth-user-pass but to be sure you should have only one instance of it.

**Comment by yoopernc** [comment]
Danke schön!

**Comment by yoopernc** [comment]
Guten Tag, @schwabe. I pulled down your new version 0.7.27, tested with it, and it works! Congratulations! Did you put this fix in an earlier version, like 0.7.25 or 0.7.26? Again, I want to thank you for responding to my situation and completing this fix so quickly. Your efforts are greatly appreciated. 👍
